File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch07/Python/P07-MatDetIdent.py
# Check matrix identity det(A*B) = det(A)*det(B)
from random import *
from linsys import *
from matutil import *

n = 100
a = [[0]*(n+1) for i in range(n+1)]
b = [[0]*(n+1) for i in range(n+1)]
c = [[0]*(n+1) for i in range(n+1)]

for i in range(1,n+1):                   # generate matrices a and b randomly
   for j in range(1,n+1):
      a[i][j] = random()
      b[i][j] = random()

MatProd(a,b,c,n,n,n)                                              # C = A * B
detA  = Gauss(a,b,n,0)                       # m=0: performs only elimination
detB  = Gauss(b,b,n,0)                       # and calculates the determinant
detAB = Gauss(c,b,n,0)

print("det(A)  = {0:9.2e}".format(detA))
print("det(B)  = {0:9.2e}".format(detB))
print("det(AB) = {0:9.2e}".format(detAB))
print("Error   = {0:9.2e}".format(1e0 - detA*detB/detAB))

File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch07/C/P07-MatDetIdent.cpp
// Check matrix identity det(A*B) = det(A)*det(B)
#include <stdio.h>
#include "memalloc.h"
#include "linsys.h"
#include "matutil.h"

int main()
{
   double **a, **b, **c, detA, detAB, detB;
   int i, j, n;

   n = 100;
   a = Matrix(1,n,1,n);
   b = Matrix(1,n,1,n);
   c = Matrix(1,n,1,n);

   for (i=1; i<=n; i++)                 // generate matrices a and b randomly
      for (j=1; j<=n; j++) {
         a[i][j] = (double)rand()/RAND_MAX;
         b[i][j] = (double)rand()/RAND_MAX;
      }

   MatProd(a,b,c,n,n,n);                                         // C = A * B
   Gauss(a,b,n,0,detA);                     // m=0: performs only elimination
   Gauss(b,b,n,0,detB);                     // and calculates the determinant
   Gauss(c,b,n,0,detAB);

   printf("det(A)  = %10.2e\n",detA);
   printf("det(B)  = %10.2e\n",detB);
   printf("det(AB) = %10.2e\n",detAB);
   printf("Error   = %10.2e\n",1e0 - detA*detB/detAB);
}
